beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java

```diff
@@ -0,0 +1,194 @@
+import cz.o2.proxima.repository.Repository;
+import cz.o2.proxima.repository.RepositoryFactory;
+import cz.o2.proxima.storage.Partition;
+import cz.o2.proxima.storage.StreamElement;
+import cz.o2.proxima.storage.commitlog.Position;
+import cz.o2.proxima.time.Watermarks;
+import cz.o2.proxima.util.ExceptionUtils;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.SerializableCoder;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Impulse;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/** A {@link PTransform} that reads from a {@link CommitLogReader} using splittable DoFn. */
+public class CommitLogRead extends PTransform<PBegin, PCollection<StreamElement>> {
+
+  /**
+   * Create the {@link CommitLogRead} transform.
+   *
+   * @param observeName name of the observer
+   * @param limit limit (use {@link Long#MAX_VALUE} for unbounded
+   * @param repo repository
+   * @param reader the reader
+   * @return {@link CommitLogRead} transform for the commit log
+   */
+  public static CommitLogRead of(
+      String observeName, long limit, Repository repo, CommitLogReader reader) {
+
+    return new CommitLogRead(observeName, limit, repo, reader);
+  }
+
+  @DoFn.UnboundedPerElement
+  private static class CommitLogReadFn extends DoFn<byte[], StreamElement> {
+
+    private final String name;
+    private final long limit;
+    private final RepositoryFactory repositoryFactory;
+    private final CommitLogReader.Factory<?> readerFactory;
+
+    private transient boolean externalizableOffsets = false;
+    private transient BlockingQueueLogObserver observer;
+
+    private CommitLogReadFn(
+        String name,
+        long limit,
+        RepositoryFactory repositoryFactory,
+        CommitLogReader.Factory<?> readerFactory) {
+
+      this.name = name;
+      this.limit = limit;
+      this.repositoryFactory = repositoryFactory;
+      this.readerFactory = readerFactory;
+    }
+
+    @ProcessElement
+    public ProcessContinuation process(
+        RestrictionTracker<OffsetRestrictionTracker.StartEndOffset, Offset> tracker,
+        OutputReceiver<StreamElement> output) {
+
+      try (ObserveHandle handle = startObserve(tracker.currentRestriction())) {
+        int i = 0;
+        while (!Thread.currentThread().isInterrupted()
+            && observer.getWatermark() < Watermarks.MAX_WATERMARK) {
+
+          StreamElement element = observer.take();
+          if (element == null) {
+            ExceptionUtils.ignoringInterrupted(() -> TimeUnit.MILLISECONDS.sleep(100));
```

Comment:
Should we return `ProcessContinuation.resume()` here instead of blocking?

Reply:
This is closely related to https://github.com/datadrivencz/proxima-platform/pull/458/files#r522286094

From FnApiDoFnRunner it seems useful to return `resume`, because we could checkpoint the work that has been already done (org.apache.beam.fn.harness.FnApiDoFnRunner#trySplitForElementAndRestriction)

Reply:
Agree, this needs addressing. Not yet sure how to solve this, but we have to.